content/technical/open-source-as-a-service.md

```diff
@@ -0,0 +1,155 @@
+are interconnected with other services, with data, APIs, legal
+agreements, process, and users, infrastructure, and so much more.
+
+Any non-toy Service is not reproducible.
+
+In reality the threat here is far more fundamental: Users of services
+cannot readily change something in the service not only because the
+processes is operated by someone else ... but because such users of
+a service have explicitly chosen not to be involved in operating
+the software.
+
+If the primary way to use Postgres was as a Service, then that (fabulous)
+project would be starved for contributors. This is because the number
+of contributors in a such a project are a function of some small
+percentage of users deciding to try to make change.
+
+As this mechanism for using software, as a service, takes over the world
+the pool of users who can contribute to Open Source shrinks dramatically.
+
+For a long time I saw this as a fundamental threat that would starve
+Open Source. I was unable reconcile Software as a Service with a healthy
+Open Source ecosystem.
+
+But recently, I've become convinced that if we enable an effective
+contribution model on a service, one that doesn't require that the users
+themselves become operators of the service ... then we can reconcile this
+threat and Open Source will thrive on services instead of be starved by them.
+
+
+# You want me to do what!?
+
+To enable contributions by users of a service, users that have explicitly
+chosen not to operate the software themselves, we have to set up a process
+by which they can make a change to a running service and experience that
+change themselves before others do.
+
+For example, a user of such a hypothetical service should be able to:
+
+ 1. Discover which component of a service to contribute to.
+
+ 2. Make a non-sensical change (like add a printf style log statement) or change the spelling of a word.
+
+ 3. Experience that change when they use the service, or when it acts on their data.
+
+After discussing this with others, I firmly believe we have all the techniques we need,
+whether canary deployments, load balancing, continuous delivery,
+infrastructure as code, and so on. It's been hard to find a single
+problem that an existing practice doesn't solve.
+
+It turns out many engineering teams working have come to that same
+conclusion. Rapid deployment of changes to services is a powerful
+capability highly sought after. For example,
+[GitHub deploys changes before they're merged](https://github.blog/2015-06-02-deploying-branches-to-github-com/).
+Whereas
+[Facebook rapidly deploys changes](https://engineering.fb.com/2017/08/31/web/rapid-release-at-massive-scale/)
+to a few canary machines and scales each change up to production.
+Everyone has their own version of continuous delivery. It's hard to
+take an engineering team seriously if they haven't figured out how
+their team members can get a change rapidly deployed.
+
+As Open Source, I don't see us having pulled this off. Not yet.
```

Comment:
This is something different, though, so IMHO this breaks the flow of argument here. This paragraph is about the engineers of the SaaS installation doing CD of their own code to their own official installation. I don't know how gitlab or e.g. the Testing Farm cluster handles that, but I can't imagine that CD is so rare in the FOSS world? (FWIW, this is exactly how we handle infra changes in our cockpit CI, even though that is not a public service).

Your plea above was that developers *outside* of that core developer team could submit a PR and get their own shard of the service (*not* get it deployed to the official primary SaaS). The latter is indeed much harder IMHO, but then again I am not aware of any non-open project which does that. In fact, by definition you couldn't even do something like "send a PR against github".

So, I agree that FOSS has not managed to set up this "get your own shard on a PR", and this is indeed something that would give FOSS infra projects a decisive advantage over their closed-source counterparts. But it's certainly not an aspect where FOSS infra projects lag behind closed-source ones. So this could be easy to msunderstand here.

Reply:
Indeed. This was sloppy of me. I removed a bunch of the "how do we do it" stuff for later posts because the article was getting too long. How does it look now?